main.cu

```diff
@@ -12,6 +18,66 @@ inline void gpuAssert(cudaError_t code, const char *file, int line, bool abort=t
    }
 }
 
+
+//=========================================
+
+//Naive Implementation of Euclidean_distance_matrix (BruteForce)
+extern "C" void matrix_Euclidean_distance_same_block_cpu(double * a, double * b, double * c, int * n, int * m, int * m_b) {
+  int array_size = * n * * m;
+  float * array_new = new float[ * n * * m];
+  for (int i = 0; i < array_size; ++i) {
+    array_new[i] = a[i];
+  }
+
+  float * d_array = new float[array_size];
+
+  std::memcpy(d_array, array_new, array_size * sizeof(float));
+
+  unsigned int * d_result = new unsigned int[( * m) * ( * m)];
+  unsigned int * h_result = new unsigned int[( * m) * ( * m)];
+  std::memset(d_result, 0, ( * m) * ( * m) * sizeof(unsigned int));
+
+  //CPU Implementation
+  for (int row = 0; row < * n; row++) {
+    //Reuclidean_cpu_atomic_float(i,d_array, *n, *m, d_result);
+
+    for (int col1_num = 0; col1_num < * m; ++col1_num) {
+      for (int col2_num = col1_num + 1; col2_num < * m; ++col2_num) {
+        float * col1 = d_array + * n * col1_num;
+        float * col2 = d_array + * n * col2_num;
+        if (row < * n) {
+          float diff = col1[row] - col2[row];
+          diff = diff * diff;
+
+          //Can't understand the logic here of atomicAdd
+          /*So far, if we are taking the address (result+col1*m +col2_num) and adding diff to it
+              atomicAdd(A,B)-> takes address of A add B to it.(inplace addition)
+          */
+          d_result[col1_num * * m + col2_num] += diff;
+          if (row < 1)
+            std::cout << d_result;
+          //d_result+= diff;
+          d_result[col2_num * * m + col1_num] += diff;
+          //atomicAdd(result + col1_num * m + col2_num, diff);
+          //atomicAdd(result + col2_num * m + col1_num, diff);
+
+        }
+      }
+    }
+  }
+
+  std::memcpy(h_result, d_result, ( * m) * ( * m) * sizeof(float));

```

Comment:
You should not perform copying memory. This is needed only for copying from CPU to GPU and vice versa